iothub_client/src/iothubtransporthttp.c

```diff
@@ -2023,111 +2023,83 @@ static bool abandonOrAcceptMessage(HTTPTRANSPORT_HANDLE_DATA* handleData, HTTPTR
             else
             {
-                if ((tc->handleData == NULL) || (tc->deviceData == NULL) || (tc->etagValue == NULL))
+                HTTPTRANSPORT_PERDEVICE_DATA* deviceData = (HTTPTRANSPORT_PERDEVICE_DATA*)handle;
+
+                if (abandonOrAcceptMessage(deviceData->transportHandle, deviceData, dispositionContext->etagValue, disposition))
                 {
-                    /* Codes_SRS_TRANSPORTMULTITHTTP_10_002: [If any of the messageData fields are NULL, IoTHubTransportHttp_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR.] */
-                    LogError("invalid transport context data");
-                    result = IOTHUB_CLIENT_ERROR;
+                    result = IOTHUB_CLIENT_OK;
                 }
                 else
                 {
-                    if (abandonOrAcceptMessage(tc->handleData, tc->deviceData, tc->etagValue, disposition))
-                    {
-                        result = IOTHUB_CLIENT_OK;
-                    }
-                    else
-                    {
-                        /* Codes_SRS_TRANSPORTMULTITHTTP_10_003: [IoTHubTransportHttp_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR if the POST message fails, otherwise return IOTHUB_CLIENT_OK.] */
-                        LogError("HTTP Transport layer failed to report %s disposition", MU_ENUM_TO_STRING(IOTHUBMESSAGE_DISPOSITION_RESULT, disposition));
-                        result = IOTHUB_CLIENT_ERROR;
-                    }
+                    /* Codes_SRS_TRANSPORTMULTITHTTP_10_003: [IoTHubTransportHttp_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR if the POST message fails, otherwise return IOTHUB_CLIENT_OK.] */
+                    LogError("HTTP Transport layer failed to report %s disposition", MU_ENUM_TO_STRING(IOTHUBMESSAGE_DISPOSITION_RESULT, disposition));
+                    result = IOTHUB_CLIENT_ERROR;
                 }
-                free(tc->etagValue);
-                free(tc);
             }
-            IoTHubMessage_Destroy(message_data->messageHandle);
         }
-        free(message_data);
+
+        // This will destroy the disposition context as well.
+        IoTHubMessage_Destroy(message);
     }
     return result;
 }
 
-static MESSAGE_CALLBACK_INFO* MESSAGE_CALLBACK_INFO_Create(IOTHUB_MESSAGE_HANDLE received_message, HTTPTRANSPORT_HANDLE_DATA* handleData, HTTPTRANSPORT_PERDEVICE_DATA* deviceData, const char* etagValue)
+static void DestroyMessageDispositionContext(MESSAGE_DISPOSITION_CONTEXT* dispositionContext)
+{
+    if (dispositionContext != NULL)
+    {
+        free(dispositionContext->etagValue);
+        free(dispositionContext);
+    }
+}
+
+static MESSAGE_DISPOSITION_CONTEXT* CreateMessageDispositionContext(const char* etagValue)
 {
-    MESSAGE_CALLBACK_INFO* result = (MESSAGE_CALLBACK_INFO*)malloc(sizeof(MESSAGE_CALLBACK_INFO));
+    MESSAGE_DISPOSITION_CONTEXT* result = malloc(sizeof(MESSAGE_DISPOSITION_CONTEXT));
```

Comment:
I recommend a `calloc` here.  I can think of two CVEs where uninited memory was root cause so even though this code looks file as-is for now, as defense in depth recommend this.

Reply:
That makes all the sense, thanks for that. Changed to calloc.